Problem deserializing a map of bing tiles

```
presto:sf100> select map(array[bing_tile(1, 2, 10)], array[bing_tile(3, 4, 11)]);
java.lang.RuntimeException: Error fetching next at http://localhost:8080/v1/statement/20180108_190323_00004_ibwu6/1 returned an invalid response: JsonResponse{statusCode=200, statusMessage=OK, headers={content-type=[application/json], date=[Mon, 08 Jan 2018 19:03:23 GMT], vary=[Accept-Encoding, User-Agent], x-content-type-options=[nosniff]}, hasValue=false} [Error: {"id":"20180108_190323_00004_ibwu6","infoUri":"http://localhost:8080/query.html?20180108_190323_00004_ibwu6","partialCancelUri":"http://127.0.0.1:8080/v1/stage/20180108_190323_00004_ibwu6.0","nextUri":"http://localhost:8080/v1/statement/20180108_190323_00004_ibwu6/2","columns":[{"name":"_col0","type":"map(BingTile,BingTile)","typeSignature":{"rawType":"map","typeArguments":[{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]},{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}],"literalArguments":[],"arguments":[{"kind":"TYPE_SIGNATURE","value":{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}},{"kind":"TYPE_SIGNATURE","value":{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}}]}}],"data":[[{"BingTile{x=1, y=2, zoom_level=10}":{"x":3,"y":4,"zoom":11}}]],"stats":{"state":"RUNNING","queued":false,"scheduled":true,"nodes":1,"totalSplits":0,"queuedSplits":0,"runningSplits":0,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"queuedTimeMillis":4,"elapsedTimeMillis":76,"processedRows":0,"processedBytes":0,"peakMemoryBytes":0,"rootStage":{"stageId":"0","state":"RUNNING","done":false,"nodes":1,"totalSplits":0,"queuedSplits":0,"runningSplits":0,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"subStages":[]}}}
]
	at com.facebook.presto.client.StatementClient.requestFailedException(StatementClient.java:390)
	at com.facebook.presto.client.StatementClient.advance(StatementClient.java:333)
	at com.facebook.presto.cli.StatusPrinter.printInitialStatusUpdates(StatusPrinter.java:122)
	at com.facebook.presto.cli.Query.renderQueryOutput(Query.java:134)
	at com.facebook.presto.cli.Query.renderOutput(Query.java:118)
	at com.facebook.presto.cli.Console.process(Console.java:298)
	at com.facebook.presto.cli.Console.runConsole(Console.java:250)
	at com.facebook.presto.cli.Console.run(Console.java:145)
	at com.facebook.presto.cli.Presto.main(Presto.java:32)
Caused by: java.lang.IllegalArgumentException: Unable to create class com.facebook.presto.client.QueryResults from JSON response:
[{"id":"20180108_190323_00004_ibwu6","infoUri":"http://localhost:8080/query.html?20180108_190323_00004_ibwu6","partialCancelUri":"http://127.0.0.1:8080/v1/stage/20180108_190323_00004_ibwu6.0","nextUri":"http://localhost:8080/v1/statement/20180108_190323_00004_ibwu6/2","columns":[{"name":"_col0","type":"map(BingTile,BingTile)","typeSignature":{"rawType":"map","typeArguments":[{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]},{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}],"literalArguments":[],"arguments":[{"kind":"TYPE_SIGNATURE","value":{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}},{"kind":"TYPE_SIGNATURE","value":{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}}]}}],"data":[[{"BingTile{x=1, y=2, zoom_level=10}":{"x":3,"y":4,"zoom":11}}]],"stats":{"state":"RUNNING","queued":false,"scheduled":true,"nodes":1,"totalSplits":0,"queuedSplits":0,"runningSplits":0,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"queuedTimeMillis":4,"elapsedTimeMillis":76,"processedRows":0,"processedBytes":0,"peakMemoryBytes":0,"rootStage":{"stageId":"0","state":"RUNNING","done":false,"nodes":1,"totalSplits":0,"queuedSplits":0,"runningSplits":0,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"subStages":[]}}}
]
	at com.facebook.presto.client.JsonResponse.<init>(JsonResponse.java:69)
	at com.facebook.presto.client.JsonResponse.execute(JsonResponse.java:143)
	at com.facebook.presto.client.StatementClient.advance(StatementClient.java:320)
	... 7 more
Caused by: java.lang.IllegalArgumentException: Invalid JSON string for [simple type, class com.facebook.presto.client.QueryResults]
	at io.airlift.json.JsonCodec.fromJson(JsonCodec.java:137)
	at com.facebook.presto.client.JsonResponse.<init>(JsonResponse.java:66)
	... 9 more
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.facebook.presto.client.QueryResults, problem: Illegal base64 character 7b
 at [Source: {"id":"20180108_190323_00004_ibwu6","infoUri":"http://localhost:8080/query.html?20180108_190323_00004_ibwu6","partialCancelUri":"http://127.0.0.1:8080/v1/stage/20180108_190323_00004_ibwu6.0","nextUri":"http://localhost:8080/v1/statement/20180108_190323_00004_ibwu6/2","columns":[{"name":"_col0","type":"map(BingTile,BingTile)","typeSignature":{"rawType":"map","typeArguments":[{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]},{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}],"literalArguments":[],"arguments":[{"kind":"TYPE_SIGNATURE","value":{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}},{"kind":"TYPE_SIGNATURE","value":{"rawType":"BingTile","typeArguments":[],"literalArguments":[],"arguments":[]}}]}}],"data":[[{"BingTile{x=1, y=2, zoom_level=10}":{"x":3,"y":4,"zoom":11}}]],"stats":{"state":"RUNNING","queued":false,"scheduled":true,"nodes":1,"totalSplits":0,"queuedSplits":0,"runningSplits":0,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"queuedTimeMillis":4,"elapsedTimeMillis":76,"processedRows":0,"processedBytes":0,"peakMemoryBytes":0,"rootStage":{"stageId":"0","state":"RUNNING","done":false,"nodes":1,"totalSplits":0,"queuedSplits":0,"runningSplits":0,"completedSplits":0,"userTimeMillis":0,"cpuTimeMillis":0,"wallTimeMillis":0,"processedRows":0,"processedBytes":0,"subStages":[]}}}
; line: 1, column: 1420]
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:268)
	at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1405)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.wrapAsJsonMappingException(StdValueInstantiator.java:468)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.rewrapCtorProblem(StdValueInstantiator.java:487)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:276)
	at com.fasterxml.jackson.databind.deser.ValueInstantiator.createFromObjectWith(ValueInstantiator.java:224)
	at com.fasterxml.jackson.databind.deser.impl.PropertyBasedCreator.build(PropertyBasedCreator.java:135)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer._deserializeUsingPropertyBased(BeanDeserializer.java:471)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.deserializeFromObjectUsingNonDefault(BeanDeserializerBase.java:1194)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserializeFromObject(BeanDeserializer.java:314)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:148)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:3789)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:2871)
	at io.airlift.json.JsonCodec.fromJson(JsonCodec.java:134)
	... 10 more
Caused by: java.lang.IllegalArgumentException: Illegal base64 character 7b
	at java.util.Base64$Decoder.decode0(Base64.java:714)
	at java.util.Base64$Decoder.decode(Base64.java:526)
	at java.util.Base64$Decoder.decode(Base64.java:549)
	at com.facebook.presto.client.QueryResults.fixValue(QueryResults.java:339)
	at com.facebook.presto.client.QueryResults.fixValue(QueryResults.java:265)
	at com.facebook.presto.client.QueryResults.fixData(QueryResults.java:237)
	at com.facebook.presto.client.QueryResults.<init>(QueryResults.java:97)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.fasterxml.jackson.databind.introspect.AnnotatedConstructor.call(AnnotatedConstructor.java:124)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromObjectWith(StdValueInstantiator.java:274)
	... 19 more
Query is gone (server restarted?)
```

--- a/presto-client/src/main/java/com/facebook/presto/client/QueryResults.java
+++ b/presto-client/src/main/java/com/facebook/presto/client/QueryResults.java
@@ -37,6 +37,7 @@ import java.util.Set;
 
 import static com.facebook.presto.spi.type.StandardTypes.ARRAY;
 import static com.facebook.presto.spi.type.StandardTypes.BIGINT;
+import static com.facebook.presto.spi.type.StandardTypes.BING_TILE;
 import static com.facebook.presto.spi.type.StandardTypes.BOOLEAN;
 import static com.facebook.presto.spi.type.StandardTypes.CHAR;
 import static com.facebook.presto.spi.type.StandardTypes.DATE;
@@ -332,6 +333,10 @@ public class QueryResults
             case CHAR:
             case GEOMETRY:
                 return String.class.cast(value);
+            case BING_TILE:
+                // Bing tiles are serialized as strings when used as map keys,
+                // they are serialized as json otherwise (value will be a LinkedHashMap).
+                return value;
             default:
                 // for now we assume that only the explicit types above are passed
                 // as a plain text and everything else is base64 encoded binary

--- a/presto-client/src/test/java/com/facebook/presto/client/TestQueryResults.java
+++ b/presto-client/src/test/java/com/facebook/presto/client/TestQueryResults.java
@@ -53,6 +53,7 @@ public class TestQueryResults
         assertQueryResult("json", "{\"json\": {\"a\": 1}}", "{\"json\": {\"a\": 1}}");
         assertQueryResult("ipaddress", "1.2.3.4", "1.2.3.4");
         assertQueryResult("Geometry", "POINT (1.2 3.4)", "POINT (1.2 3.4)");
+        assertQueryResult("map(BingTile,bigint)", ImmutableMap.of("BingTile{x=1, y=2, zoom_level=10}", 1), ImmutableMap.of("BingTile{x=1, y=2, zoom_level=10}", 1L));
     }
 
     private void assertQueryResult(String type, Object data, Object expected)

--- a/presto-spi/src/main/java/com/facebook/presto/spi/type/StandardTypes.java
+++ b/presto-spi/src/main/java/com/facebook/presto/spi/type/StandardTypes.java
@@ -41,6 +41,7 @@ public final class StandardTypes
     public static final String JSON = "json";
     public static final String IPADDRESS = "ipaddress";
     public static final String GEOMETRY = "Geometry";
+    public static final String BING_TILE = "BingTile";
 
     private StandardTypes() {}
 }

